File: javascript/BookManager/test/bookManager.test.js
const BookManager = require('../src/bookManager')
const Book = require('../src/book')

describe('BookManager', () => {
    let bookManager

    beforeEach(() => {
        bookManager = new BookManager()
    })

    afterEach(() => {
        bookManager = null
    })

    // book class
    test('Book constructor should create book instance with valid parameters', () => {
        const book = new Book("Cara membedakan kucing dan ayam", "Rama", 2025)
        expect(book.title).toBe("Cara membedakan kucing dan ayam")
        expect(book.author).toBe("Rama")
        expect(book.year).toBe(2025)
    })

    test('Book constructor should throw error when parameters is not valid', () => {
        expect(() => new Book(null, "Rama", 2025)).toThrow("Title cannot be null")
        expect(() => new Book("  ", "Rama", 2025)).toThrow("Title cannot be null")
        expect(() => new Book("Cara membedakan kucing dan ayam", null, 2025)).toThrow("Author cannot be null")
        expect(() => new Book("Cara membedakan kucing dan ayam", "  ", 2025)).toThrow("Author cannot be null")
        expect(() => new Book("Cara membedakan kucing dan ayam", "Rama", -1)).toThrow("Year must between 0 and 2100")
        expect(() => new Book("Cara membedakan kucing dan ayam", "Rama", 2101)).toThrow("Year must between 0 and 2100")
    })

    test('Book equals should return true for same book details', () => {
        const book1 = new Book("Cara membedakan kucing dan ayam", "Rama", 2025)
        const book2 = new Book("Cara membedakan kucing dan ayam", "Rama", 2025)
        expect(book1.equals(book2)).toBe(true)
    })

    test('Book equals should return false for different book details', () => {
        const book1 = new Book("Cara membedakan kucing dan ayam", "Rama", 2025)
        const book2 = new Book("Cara membedakan kaki kanan dan kiri", "Rina", 2025)
        expect(book1.equals(book2)).toBe(false)
    })

    test('Book equals should return false when comparing with non-book object', () => {
        const book = new Book("Cara membedakan kucing dan ayam", "Rama", 2025)
        expect(book.equals(null)).toBe(false)
        expect(book.equals({})).toBe(false)
    })

    test('Book toString should return formatted string', () => {
        const book = new Book("Cara membedakan kucing dan ayam", "Rama", 2025)
        expect(book.toString()).toBe("Cara membedakan kucing dan ayam by Rama (2025)")
    })
    // book class

    // toLower function
    test('toLower should convert string to lowercase', () => {
        expect(bookManager.toLower("RAMA")).toBe("rama");
        expect(bookManager.toLower("TeSt")).toBe("test");
        expect(bookManager.toLower("")).toBe("");
    });

    test('toLower should return param as is if not a string', () => {
        expect(bookManager.toLower(123)).toBe(123);
        expect(bookManager.toLower(null)).toBe(null);
        expect(bookManager.toLower(undefined)).toBe(undefined);
    });
    // toLower function

    // addbook function
    test('addBook should add book to the collection', () => {
        const book = new Book("Cara membedakan kucing dan ayam", "Rama", 2025)
        bookManager.addBook(book)
        expect(bookManager.getBookCount()).toBe(1)
    })

    test('addBook should throw error if book is null', () => {
        expect(() => bookManager.addBook(null)).toThrow("Book cannot empty")
    })
    // addbook function

    // removeBook function
    test('removeBook should remove book by title', () => {
        const book1 = new Book("Cara membedakan kucing dan ayam", "Rama", 2025)
        const book2 = new Book("Cara membedakan kaki kanan dan kiri", "Rina", 2025)
        bookManager.addBook(book1)
        bookManager.addBook(book2)
        const removed = bookManager.removeBook("Cara membedakan kucing dan ayam")
        expect(removed).toBe(true)
        expect(bookManager.getBookCount()).toBe(1)
        expect(bookManager.getAllBooks()).toEqual([book2])
    })

    // lengkapi test unit untuk removeBook jika buku tidak ditemukan
    test('removeBook should return false if book title not found', () => {
        const book = new Book("Cara membedakan kucing dan ayam", "Rama", 2025)
        bookManager.addBook(book)
        const removed = bookManager.removeBook("Non-existing title")
        expect(removed).toBe(false)
        expect(bookManager.getBookCount()).toBe(1)
    })

    test('removeBook should throw error if title is null or empty', () => {
        expect(() => bookManager.removeBook(null)).toThrow("Title cannot be null")
    })
    // removeBook function

    // findBookByAuthor function
    // lengkapi test unit untuk mencari buku berdasarkan author
    test('findBookByAuthor should return books by the specified author', () => {
        const book1 = new Book("Cara membedakan kucing dan ayam", "Rama", 2025)
        const book2 = new Book("Cara membedakan kaki kanan dan kiri", "Rina", 2025)
        const book3 = new Book("Cara merawat kucing", "Rama", 2024)
        bookManager.addBook(book1)
        bookManager.addBook(book2)
        bookManager.addBook(book3)
        const results = bookManager.findBooksByAuthor("Rama")
        expect(results.length).toBe(2)
        expect(results).toEqual([book1, book3])
    })

    test('findBookByAuthor should throw error if author is null or empty', () => {
        expect(() => bookManager.findBooksByAuthor(null)).toThrow("Author cannot be null")
    })
    // findBookByAuthor function

    // getAllBooks function
    // lengkapi test unit untuk seluruh buku yang ada di list
    test('getAllBooks should return all books in the collection', () => {
        const book1 = new Book("Cara membedakan kucing dan ayam", "Rama", 2025)
        const book2 = new Book("Cara membedakan kaki kanan dan kiri", "Rina", 2025)
        bookManager.addBook(book1)
        bookManager.addBook(book2)
        const allBooks = bookManager.getAllBooks()
        expect(allBooks.length).toBe(2)
        expect(allBooks).toEqual([book1, book2])
    })
    // getAllBooks function

    // clearBooks function
    test('clearBooks should remove all books from the collection', () => {
        const book1 = new Book("Cara membedakan kucing dan ayam", "Rama", 2025)
        const book2 = new Book("Cara membedakan kaki kanan dan kiri", "Rina", 2025)
        bookManager.addBook(book1)
        bookManager.addBook(book2)
        bookManager.clearAllBooks()
        expect(bookManager.getBookCount()).toBe(0)
        expect(bookManager.getAllBooks()).toEqual([])
    })
    // clearBooks function


    // findBookByYear function
    test('findBookByYear should return books published in the specified year', () => {
        const book1 = new Book("Cara membedakan kucing dan ayam", "Rama", 2025)
        const book2 = new Book("Cara membedakan kaki kanan dan kiri", "Rina", 2024)
        const book3 = new Book("Cara merawat kucing", "Rama", 2025)
        bookManager.addBook(book1)
        bookManager.addBook(book2)
        bookManager.addBook(book3)
        const results = bookManager.findBookByYear(2025)
        expect(results.length).toBe(2)
        expect(results).toEqual([book1, book3])
    })

    test('findBookByYear should throw error if year is out of range', () => {
        expect(() => bookManager.findBookByYear(-1)).toThrow("Year must between 0 and 2100")
        expect(() => bookManager.findBookByYear(2200)).toThrow("Year must between 0 and 2100")
    })
    // findBookByYear function

    // containsBookTitle function
    test('containsBookTitle should return true if book title exists', () => {
        const book = new Book("Cara membedakan kucing dan ayam", "Rama", 2025)
        bookManager.addBook(book)
        expect(bookManager.containsBookTitle("Cara membedakan kucing dan ayam")).toBe(true)
    })

    test('containsBookTitle should return false if book title does not exist', () => {
        const book = new Book("Cara membedakan kucing dan ayam", "Rama", 2025)
        bookManager.addBook(book)
        expect(bookManager.containsBookTitle("Non-existing title")).toBe(false)
    })

    test('containsBookTitle should throw error if title is null or empty', () => {
        expect(() => bookManager.containsBookTitle(null)).toThrow("Title cannot be null")
    })
    // containsBookTitle function
})

File: java/BookManager/src/test/java/org/example/BookManagerTest.java
package org.example;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;


class BookManagerTest {
    private BookManager bookManager;

    @BeforeEach
    void setUp(){
        bookManager = new BookManager();
    }

    // class book
    @Test
    @DisplayName("Test book initialization")
    void testBookInitialization(){
        Book book = new Book("Cara membedakan kucing dan ayam", "Rama", 2025);
        assertEquals("Cara membedakan kucing dan ayam", book.getBook_Title());
        assertEquals("Rama", book.getAuthor());
        assertEquals(2025, book.getYears());
    }

    @Test
    @DisplayName("Test book invalid initialization")
    void testBookInvalidInitialization(){
        assertThrows(IllegalArgumentException.class, () -> new Book("  ", "Rama", 2025));
        assertThrows(IllegalArgumentException.class, () -> new Book(null, "Rama", 2025));
        assertThrows(IllegalArgumentException.class, () -> new Book("Cara membedakan kucing dan ayam", "   ", 2025));
        assertThrows(IllegalArgumentException.class, () -> new Book("Cara membedakan kucing dan ayam", null, 2025));
        assertThrows(IllegalArgumentException.class, () -> new Book("Cara membedakan kucing dan ayam", "Rama", -1));
        assertThrows(IllegalArgumentException.class, () -> new Book("Cara membedakan kucing dan ayam", "Rama", 2101));
    }

    @Test
    @DisplayName("Test book string representation")
    void testBookStringRepresentation(){
        Book book = new Book("Cara membedakan kucing dan ayam", "Rama", 2025);
        assertEquals("Cara membedakan kucing dan ayam by Rama (2025)", book.toString());
    }

    @SuppressWarnings("AssertBetweenInconvertibleTypes")
    @Test
    @DisplayName("Test book equality")
    void testBookEquality(){
        Book book1 = new Book("Cara membedakan kucing dan ayam", "Rama", 2025);
        Book book2 = new Book("Cara membedakan kucing dan ayam", "Rama", 2025);
        Book book3 = new Book("Cara membedakan kaki kanan dan kaki kiri", "Rina", 2025);
        assertEquals(book1, book2);
        assertNotEquals(book1, book3);
        assertNotEquals("this is not book", book1);
        assertNotEquals(null, book1);
    }

    @Test
    @DisplayName("Test book object hash")
    void testBookHash(){
        Book book1 = new Book("Cara membedakan kucing dan ayam", "Rama", 2025);
        Book book2 = new Book("Cara membedakan kucing dan ayam", "Rama", 2025);
        Book book3 = new Book("Cara membedakan kaki kanan dan kaki kiri", "Rina", 2025);
        assertEquals(book1.hashCode(), book2.hashCode());
        assertNotEquals(book2.hashCode(), book3.hashCode());
    }
    // class book

    // addBook function
    @Test
    @DisplayName("Test add book")
    void testAddBook(){
        Book book = new Book("Cara membedakan kucing dan ayam", "Rama", 2025);
        bookManager.addBook(book);
        assertEquals(1, bookManager.getBookSize(),"After add some book, must return 1 of record data");
    }

    @Test
    @DisplayName("Test add with empty book")
    void testAddEmptyBook(){
        Exception exception = assertThrows(IllegalArgumentException.class, () -> bookManager.addBook(null));
        assertEquals("Book can't empty", exception.getMessage(), "When add a empty book should return IllegalArgumentException");
    }

    @Test
    @DisplayName("Test add duplicate book")
    void testAddDuplicateBook(){
        Book book1 = new Book("Cara membedakan kucing dan ayam", "Rama", 2025);
        Book book2 = new Book("Cara membedakan kucing dan ayam", "Rama", 2025);
        bookManager.addBook(book1);

        Exception exception = assertThrows(IllegalArgumentException.class, () -> bookManager.addBook(book2));

        assertEquals("Title already exist", exception.getMessage(), "");
    }

    @Test
    @DisplayName("Book with null title throws exception")
    void testBookWithNullTitle() {
        Exception ex = assertThrows(IllegalArgumentException.class, () -> new Book(null, "Rama", 2025));
        assertEquals("Title required", ex.getMessage());
    }

    @Test
    @DisplayName("Book with empty author throws exception")
    void testBookWithEmptyAuthor() {
        Exception ex = assertThrows(IllegalArgumentException.class, () -> new Book("Belajar Java", "   ", 2025));
        assertEquals("Author required", ex.getMessage());
    }

    @Test
    @DisplayName("Book with invalid year throws exception")
    void testBookWithInvalidYear() {
        Exception ex = assertThrows(IllegalArgumentException.class, () -> new Book("Belajar Java", "Rama", 1999));
        assertEquals("Years must between 2000 and 2100", ex.getMessage());
    }
    // addBook function

    // clear book
    @Test
    @DisplayName("test clear all books")
    void testClearBook(){
        Book book = new Book("Cara membedakan kucing dan ayam", "Rama", 2025);
        bookManager.addBook(book);
        assertEquals(1, bookManager.getBookSize(), "After add book must return 1 record data");
        bookManager.clearAllBooks();
        assertEquals(0, bookManager.getBookSize(), "After clear book must return 0 record data");
    }
    // clear book

    // find book by year
    @Test
    @DisplayName("Test find book by year")
    void testFindBookByYear(){
        Book book = new Book("Cara membedakan kucing dan ayam", "Rama", 2025);
        bookManager.addBook(book);
        assertEquals(1, bookManager.getBookSize(), "After add book must return 1 record data");
        assertEquals(1, bookManager.findBookByYear(2025).size());
    }

    @Test
    @DisplayName("Test find book by year without range < 2000 and > 2100")
    void testFindBookByYearWithoutYear(){
        Exception ex1 = assertThrows(IllegalArgumentException.class, () -> bookManager.findBookByYear(2101));
        Exception ex2 = assertThrows(IllegalArgumentException.class, () -> bookManager.findBookByYear(2));
        assertEquals("Years must between 2000 and 2100", ex1.getMessage());
        assertEquals("Years must between 2000 and 2100", ex2.getMessage());
    }
    // find book by year

    // check contains book
    @Test
    @DisplayName("Test check contains book with empty title")
    void testCheckContainsBookWithoutTitle(){
        Exception exNull = assertThrows(IllegalArgumentException.class, () -> bookManager.containsBook(null));
        Exception exEmpty = assertThrows(IllegalArgumentException.class, () -> bookManager.containsBook("  "));
        assertEquals("Title can't empty", exNull.getMessage());
        assertEquals("Title can't empty", exEmpty.getMessage());
    }
    // check contains book

    //Lengkapi Unit Test dibawah untuk buku yang memang tidak terdapat pada list
    @Test
    @DisplayName("Test menghapus buku yang tidak ada")
    void testRemoveNonExistingBook() {
       assertFalse(bookManager.removeBook("Cara membedakan kucing dan ayam"));
    }
    // additional
    @Test
    @DisplayName("Test remove existing book")
    void testRemoveExistingBook(){
        Book book = new Book("Cara membedakan kucing dan ayam", "Rama", 2025);
        bookManager.addBook(book);
        assertEquals(1, bookManager.getBookSize(), "After add book must return 1 record data");
        bookManager.removeBook("Cara membedakan kucing dan ayam");
        assertEquals(0, bookManager.getBookSize(), "After clear book must return 0 record data");
    }

    @Test
    @DisplayName("Test remove book with no title")
    void testRemoveBookWithoutTitle(){
        Exception exNull = assertThrows(IllegalArgumentException.class, () -> bookManager.removeBook(null));

        Exception exEmpty = assertThrows(IllegalArgumentException.class, () -> bookManager.removeBook(" "));
        assertEquals("Title can't empty", exNull.getMessage());
        assertEquals("Title can't empty", exEmpty.getMessage());
    }

    //Lengkapi Unit Test dibawah untuk bmencari buku berdasarkan penulis
    @Test
    @DisplayName("Test mencari buku berdasarkan author")
    void testFindBooksByAuthor() {
        Book book1 = new Book("Cara membedakan kucing dan ayam", "Rama", 2025);
        Book book2 = new Book("Cara membedakan kaki kiri dan kanan", "Rama", 2025);
        Book book3 = new Book("Cara membuka kulkas", "Rina", 2025);

        bookManager.addBook(book1);
        bookManager.addBook(book2);
        bookManager.addBook(book3);
        assertEquals(3, bookManager.getBookSize());
        assertEquals(1,bookManager.findBookByAuthor("Rina").size());
        assertEquals(2,bookManager.findBookByAuthor("Rama").size());
    }
    // additional
    @Test
    @DisplayName("Find book with null author")
    void testFindBookWithNullAuthor(){
        Exception exNull = assertThrows(IllegalArgumentException.class, () -> bookManager.findBookByAuthor(null));

        Exception exEmpty = assertThrows(IllegalArgumentException.class, () -> bookManager.findBookByAuthor(" "));
        assertEquals("Author can't empty", exNull.getMessage());
        assertEquals("Author can't empty", exEmpty.getMessage());
    }

    //Lengkapi Unit Test dibawah untuk seluruh buku yang ada di dalam list
    @Test
    @DisplayName("Test mendapatkan semua buku")
    void testGetAllBooks() {
        Book book1 = new Book("Cara membedakan kucing dan ayam", "Rama", 2025);
        Book book2 = new Book("Cara membedakan kaki kiri dan kanan", "Rama", 2025);
        Book book3 = new Book("Cara membuka kulkas", "Rina", 2025);

        bookManager.addBook(book1);
        bookManager.addBook(book2);
        bookManager.addBook(book3);

        assertEquals(3, bookManager.getAllBooks().size());
    }
}